In [0]:
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
import pandas as pd
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
link="https://drive.google.com/open?id=1bgxgK8wkdbl2WVvipCOuB4xe3vd7rp4Y"
fluff, id = link.split('=')
downloaded = drive.CreateFile({'id':id})
downloaded.GetContentFile('500Cities_cleandata_ReadyForEDA(1).csv')

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import numpy as np
import pandas as pd
from pandas import read_csv
from pandas.plotting import scatter_matrix
from matplotlib import pyplot
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from numpy import arange
from matplotlib import pyplot
from pandas import read_csv
from pandas import set_option
from pandas.plotting import scatter_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.metrics import mean_squared_error

Q1: Intro

In [0]:
df = pd.read_csv('/content/drive/My Drive/OPIM 5512/Assignment 2/500Cities_cleandata_ReadyForEDA(1).csv')

In [0]:
df=df.dropna()#prep

In [9]:
df.shape

(453, 34)

In [10]:
df.head()

Unnamed: 0 StateAbbr  ... TEETHLOST_CrudePrev                       Geolocation
0           1        CA  ...                 6.8  (38.67504943280, -121.147605753)
1           2        FL  ...                18.3  (27.90909077340, -82.7714203383)
2           3        CA  ...                 6.7  (37.87256787650, -122.274907975)
3           4        CA  ...                11.2  (38.29804246490, -122.301093331)
4           5        FL  ...                16.2  (26.15468783030, -80.2998411020)

[5 rows x 34 columns]

In [0]:
df = df.drop(['Geolocation','StateAbbr','PlaceName', 'PlaceFIPS', 'Unnamed: 0'], axis=1)# X, covariates: Population2010 and all other health variables

In [12]:
df.head()

Population2010  ACCESS2_CrudePrev  ...  STROKE_CrudePrev  TEETHLOST_CrudePrev
0           72203                7.5  ...               1.9                  6.8
1           77648               19.6  ...               4.5                 18.3
2          112580                7.7  ...               1.9                  6.7
3           76915               12.3  ...               2.8                 11.2
4           84439               22.8  ...               3.7                 16.2

[5 rows x 29 columns]

In [0]:
seed = 123 #SEED

In [0]:
from sklearn.utils import shuffle
df = shuffle(df)#SHUFFLE

In [0]:
median=df['Population2010'].median()
#easy median code

In [0]:
df.loc[(df['Population2010'] > median ), 'Population2010Binary'] = "1"
df.loc[(df['Population2010'] < median ), 'Population2010Binary'] = "0"
#making the new columns with the binary based on conditions

In [17]:
df.head()

Population2010  ...  Population2010Binary
32           335709  ...                     1
465           69135  ...                     0
367           78958  ...                     0
487           81557  ...                     0
396          601222  ...                     1

[5 rows x 30 columns]

Q3: Pipelines (100 points)

In [0]:
df['Population2010Binary'] = df['Population2010Binary'].astype('float')

In [0]:
df=df.dropna()#prep

In [20]:
df.shape

(452, 30)

In [0]:
X = df.drop(['Population2010', 'Population2010Binary'], axis=1)
Y = df['Population2010Binary']

In [0]:
array = df.values
X = X.dropna()
Y = Y.dropna()
validation_size = 0.20
seed = 123
X_train, X_validation, Y_train, Y_validation = train_test_split(X, Y, test_size=validation_size, random_state=seed)

# Evaluate Algorithms
# Test options and evaluation metric
num_folds = 20
seed = 123
scoring = 'neg_mean_squared_error'

In [29]:
#instead of Spotcheck, use the Pipeline example
pipelines = []
pipelines.append(('ScaledBOOST', Pipeline([('Scaler', StandardScaler()),('BOOST', GradientBoostingClassifier())])))
pipelines.append(('ScaledCCLASS', Pipeline([('Scaler', StandardScaler()),('CCLASS', DecisionTreeClassifier())])))
pipelines.append(('ScaledFOREST', Pipeline([('Scaler', StandardScaler()),('FOREST', RandomForestClassifier())])))
pipelines.append(('ScaledLDA', Pipeline([('Scaler', StandardScaler()),('LDA', LinearDiscriminantAnalysis())])))
pipelines.append(('ScaledLOG', Pipeline([('Scaler', StandardScaler()),('LOG', LogisticRegression(max_iter=1000000))])))
pipelines.append(('ScaledKNNCLASS', Pipeline([('Scaler', StandardScaler()),('KNNCLASS', KNeighborsClassifier())])))
pipelines.append(('ScaledGAUS', Pipeline([('Scaler', StandardScaler()),('GAUS', GaussianNB())])))
pipelines.append(('ScaledETREES', Pipeline([('Scaler', StandardScaler()),('ETREES', ExtraTreesClassifier())])))
pipelines.append(('ScaledBAGS', Pipeline([('Scaler', StandardScaler()),('BAGS', BaggingClassifier())])))
results = []
names = []
for name, model in pipelines:
	kfold = KFold(n_splits=num_folds, random_state=seed,
               shuffle=True)
	cv_results = cross_val_score(model, X_train, Y_train, 
                              cv=kfold, 
                              scoring='accuracy')
	results.append(cv_results)
	names.append(name)
	msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
	print(msg)

ScaledBOOST: 0.584503 (0.138748)
ScaledCCLASS: 0.579094 (0.128912)
ScaledFOREST: 0.545760 (0.131852)
ScaledLDA: 0.559795 (0.113213)
ScaledLOG: 0.551316 (0.117300)
ScaledKNNCLASS: 0.515058 (0.114227)
ScaledGAUS: 0.584503 (0.105955)
ScaledETREES: 0.529532 (0.134347)
ScaledBAGS: 0.556579 (0.135438)


ScaledBOOST: 0.584503 (0.138748)
ScaledCCLASS: 0.579094 (0.128912)
ScaledFOREST: 0.545760 (0.131852)
ScaledLDA: 0.559795 (0.113213)
ScaledLOG: 0.551316 (0.117300)
ScaledKNNCLASS: 0.515058 (0.114227)
ScaledGAUS: 0.584503 (0.105955)
ScaledETREES: 0.529532 (0.134347)
ScaledBAGS: 0.556579 (0.135438)

So in regards to accuracy, the three that work best are BOOST, GAUS, and CCLASS. I ended up getting the gaussian to work somehow by having no parameters for it in the pipeline.**HOWEVER, the GAUS model at a time did not run at all for me and always failed when it was in the pipeline due to restrictions with priors. Knowing this, I decided to take the next best score and used LDA.** Below you will see two pipelines.

In [0]:
from sklearn.externals import joblib

In [52]:
#PIPELINES FOR BOOST, DECISION TREE, AND GAUSNB
#data
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=validation_size, random_state=seed)

#Build the pipeline
pipe_boost = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', GradientBoostingClassifier(random_state=seed))])

pipe_cclass = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', DecisionTreeClassifier(random_state=seed))])

pipe_gaus = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', GaussianNB(priors=None))])

param_range = [1, 2, 3, 4]

# Set grid search params
grid_params_boost = [{'clf__criterion': ['friedman_mse'],
		'clf__min_samples_leaf': param_range,
		'clf__max_depth': param_range,
		'clf__min_samples_split': param_range[1:]}]

grid_params_cclass = [{'clf__criterion': ['gini', 'entropy'],
		'clf__min_samples_leaf': param_range,
		'clf__max_depth': param_range,
		'clf__min_samples_split': param_range[1:]}]

grid_params_gaus = [{}]

#construct grid search(when I first ran it kfold had to be 20 but now it takes kfold. Don't really know why.)
# Construct grid searches
jobs = -1

gs_boost = GridSearchCV(estimator=pipe_boost,
			param_grid=grid_params_boost,
			scoring='accuracy',
			cv=kfold,
      n_jobs=jobs)
			
gs_cclass = GridSearchCV(estimator=pipe_cclass,
			param_grid=grid_params_cclass,
			scoring='accuracy',
			cv=kfold, 
			n_jobs=jobs)

gs_gaus = GridSearchCV(estimator=pipe_gaus,
			param_grid=grid_params_gaus,
			scoring='accuracy',
			cv=kfold)

# List of pipelines for ease of iteration
grids = [gs_boost, gs_cclass, gs_gaus]

# Dictionary of pipelines and classifier types for ease of reference
grid_dict = {0: 'Gradient Boosting Classifier', 1: 'Decision Tree Classifier',
             2: 'GaussianNB'}

# Fit the grid search objects
print('Performing model optimizations...')
best_acc = 0.0
best_clf = 0
best_gs = ''
for idx, gs in enumerate(grids):
	print('\nEstimator: %s' % grid_dict[idx])	
	# Fit grid search	
	gs.fit(X_train, Y_train)
	# Best params
	print('Best params: %s' % gs.best_params_)
	# Best training data accuracy
	print('Best training accuracy: %.3f' % gs.best_score_)
	# Predict on test data with best params
	y_pred = gs.predict(X_test)
	# Test data accuracy of model with best params
	print('Test set accuracy score for best params: %.3f ' % accuracy_score(Y_test, y_pred))
 	# Track best (highest test accuracy) model
	if accuracy_score(Y_test, y_pred) > best_acc:
		best_acc = accuracy_score(Y_test, y_pred)
		best_gs = gs
		best_clf = idx
print('\nClassifier with best test set accuracy: %s' % grid_dict[best_clf])

Performing model optimizations...

Estimator: Gradient Boosting Classifier
Best params: {'clf__criterion': 'friedman_mse', 'clf__max_depth': 1, 'clf__min_samples_leaf': 1, 'clf__min_samples_split': 2}
Best training accuracy: 0.565
Test set accuracy score for best params: 0.593 

Estimator: Decision Tree Classifier
Best params: {'clf__criterion': 'entropy', 'clf__max_depth': 3, 'clf__min_samples_leaf': 4, 'clf__min_samples_split': 2}
Best training accuracy: 0.557
Test set accuracy score for best params: 0.615 

Estimator: GaussianNB
Best params: {}
Best training accuracy: 0.576
Test set accuracy score for best params: 0.571 

Classifier with best test set accuracy: Decision Tree Classifier


Performing model optimizations...

Estimator: Gradient Boosting Classifier
Best params: {'clf__criterion': 'friedman_mse', 'clf__max_depth': 1, 'clf__min_samples_leaf': 1, 'clf__min_samples_split': 2}
Best training accuracy: 0.565
Test set accuracy score for best params: 0.593 

Estimator: Decision Tree Classifier
Best params: {'clf__criterion': 'entropy', 'clf__max_depth': 3, 'clf__min_samples_leaf': 4, 'clf__min_samples_split': 2}
Best training accuracy: 0.557
Test set accuracy score for best params: 0.615 

Estimator: GaussianNB
Best params: {}
Best training accuracy: 0.576
Test set accuracy score for best params: 0.571 

Classifier with best test set accuracy: Decision Tree Classifier

In [57]:
#Use the best fitting model and architecture to predict the holdout data (re-train on all training data, then apply to holdout model).
Model1 = DecisionTreeClassifier(criterion='entropy', max_depth=3, min_samples_leaf=4, min_samples_split=2,random_state=seed)
Model1.fit(X_train, Y_train)
predictions = Model1.predict(X_test)
print(accuracy_score(Y_test, predictions))
print(confusion_matrix(Y_test, predictions))
print(classification_report(Y_test, predictions))

0.5714285714285714
[[29  8]
 [31 23]]
              precision    recall  f1-score   support

         0.0       0.48      0.78      0.60        37
         1.0       0.74      0.43      0.54        54

    accuracy                           0.57        91
   macro avg       0.61      0.60      0.57        91
weighted avg       0.64      0.57      0.56        91



The results do not improve over the previous section, but this is expected as the model is not supposed to do this. If it were it would mean that it is overfit, and would thus not be useful when applied to new data as it is tuned to any irrgularities that may be contained within this particular training set. Also it is not using PCA or scalar.

In [49]:
#PIPELINES FOR BOOST, LDA, AND CCLASS. DECISION TREE WAS ALSO BEST AND HAD SAME CLFs SO NO NEED TO RE-TRAIN HOLDOUT
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=validation_size, random_state=seed)

#Build the pipeline
pipe_boost = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', GradientBoostingClassifier(random_state=seed))])

pipe_lda = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', LinearDiscriminantAnalysis())])

pipe_cclass = Pipeline([('scl', StandardScaler()),
			('pca', PCA(n_components=2)),
			('clf', DecisionTreeClassifier(random_state=seed))])

param_range = [1, 2, 3, 4]

# Set grid search params
grid_params_boost = [{'clf__criterion': ['friedman_mse'],
		'clf__min_samples_leaf': param_range,
		'clf__max_depth': param_range,
		'clf__min_samples_split': param_range[1:]}]

grid_params_lda = [{'clf__solver': ['lsqr', 'eigen'],
		'clf__shrinkage': ['auto'] }]

grid_params_cclass = [{'clf__criterion': ['gini', 'entropy'],
		'clf__min_samples_leaf': param_range,
		'clf__max_depth': param_range,
		'clf__min_samples_split': param_range[1:]}]

#construct grid search(when I first ran it kfold had to be 20 but now it takes kfold. Don't really know why.)
# Construct grid searches
jobs = -1

gs_boost = GridSearchCV(estimator=pipe_boost,
			param_grid=grid_params_boost,
			scoring='accuracy',
			cv=kfold,
      n_jobs=jobs) 
			
gs_lda = GridSearchCV(estimator=pipe_lda,
			param_grid=grid_params_lda,
			scoring='accuracy',
      cv=kfold,
      n_jobs=jobs)
			
gs_cclass = GridSearchCV(estimator=pipe_cclass,
			param_grid=grid_params_cclass,
			scoring='accuracy',
			cv=kfold, 
			n_jobs=jobs)

# List of pipelines for ease of iteration
grids = [gs_boost, gs_lda, gs_cclass]

# Dictionary of pipelines and classifier types for ease of reference
grid_dict = {0: 'Gradient Boosting Classifier', 1: 'Linear Discriminant Analysis', 
		2: 'Decision Tree Classifier'}

# Fit the grid search objects
print('Performing model optimizations...')
best_acc = 0.0
best_clf = 0
best_gs = ''
for idx, gs in enumerate(grids):
	print('\nEstimator: %s' % grid_dict[idx])	
	# Fit grid search	
	gs.fit(X_train, Y_train)
	# Best params
	print('Best params: %s' % gs.best_params_)
	# Best training data accuracy
	print('Best training accuracy: %.3f' % gs.best_score_)
	# Predict on test data with best params
	y_pred = gs.predict(X_test)
	# Test data accuracy of model with best params
	print('Test set accuracy score for best params: %.3f ' % accuracy_score(Y_test, y_pred))
 	# Track best (highest test accuracy) model
	if accuracy_score(Y_test, y_pred) > best_acc:
		best_acc = accuracy_score(Y_test, y_pred)
		best_gs = gs
		best_clf = idx
print('\nClassifier with best test set accuracy: %s' % grid_dict[best_clf])

Performing model optimizations...

Estimator: Gradient Boosting Classifier
Best params: {'clf__criterion': 'friedman_mse', 'clf__max_depth': 1, 'clf__min_samples_leaf': 1, 'clf__min_samples_split': 2}
Best training accuracy: 0.565
Test set accuracy score for best params: 0.593 

Estimator: Linear Discriminant Analysis
Best params: {'clf__shrinkage': 'auto', 'clf__solver': 'lsqr'}
Best training accuracy: 0.546
Test set accuracy score for best params: 0.516 

Estimator: Decision Tree Classifier
Best params: {'clf__criterion': 'entropy', 'clf__max_depth': 3, 'clf__min_samples_leaf': 4, 'clf__min_samples_split': 2}
Best training accuracy: 0.557
Test set accuracy score for best params: 0.615 

Classifier with best test set accuracy: Decision Tree Classifier


Performing model optimizations...

Estimator: Gradient Boosting Classifier
Best params: {'clf__criterion': 'friedman_mse', 'clf__max_depth': 1, 'clf__min_samples_leaf': 1, 'clf__min_samples_split': 2}
Best training accuracy: 0.565
Test set accuracy score for best params: 0.593 

Estimator: Linear Discriminant Analysis
Best params: {'clf__shrinkage': 'auto', 'clf__solver': 'lsqr'}
Best training accuracy: 0.546
Test set accuracy score for best params: 0.516 

Estimator: Decision Tree Classifier
Best params: {'clf__criterion': 'entropy', 'clf__max_depth': 3, 'clf__min_samples_leaf': 4, 'clf__min_samples_split': 2}
Best training accuracy: 0.557
Test set accuracy score for best params: 0.615 

Classifier with best test set accuracy: Decision Tree Classifier